File: public/packages/dotnet/caching/distributed-redis/JsonCacheSerializer.cs
// -----------------------------------------------------------------------
// <copyright file="JsonCacheSerializer.cs" company="DCSV">
// Copyright (c) DCSV. Licensed under the Apache License, Version 2.0.
// </copyright>
// -----------------------------------------------------------------------

namespace DcsvIo.D2.Caching.Distributed.Redis;

using System.Text.Json;
using DcsvIo.D2.I18n;
using DcsvIo.D2.Result;

/// <summary>
/// Default <see cref="ICacheSerializer"/> using <c>System.Text.Json</c>.
/// Dev-friendly (Redis CLI can inspect values directly), no extra deps.
/// Swap to a binary serializer (MessagePack, Protobuf) if size becomes
/// the dominant cost.
/// </summary>
public sealed class JsonCacheSerializer : ICacheSerializer
{
    private static readonly JsonSerializerOptions sr_options = new(JsonSerializerDefaults.Web)
    {
        WriteIndented = false,
        PropertyNamingPolicy = JsonNamingPolicy.CamelCase,
        ReferenceHandler = System.Text.Json.Serialization.ReferenceHandler.IgnoreCycles,
    };

    /// <inheritdoc />
    public string ContentType => "application/json";

    /// <inheritdoc />
    public D2Result<byte[]> Serialize<T>(T value)
    {
        try
        {
            var bytes = JsonSerializer.SerializeToUtf8Bytes(value, sr_options);
            return D2Result<byte[]>.Ok(bytes);
        }
        catch (JsonException)
        {
            return new D2Result<byte[]>(
                success: false,
                data: default,
                messages: [TK.Common.Errors.COULD_NOT_BE_SERIALIZED],
                statusCode: System.Net.HttpStatusCode.InternalServerError,
                errorCode: ErrorCodes.COULD_NOT_BE_SERIALIZED);
        }
    }

    /// <inheritdoc />
    public D2Result<T> Deserialize<T>(byte[] bytes)
    {
        try
        {
            // Note: a null result is a LEGITIMATE round-trip for nullable T
            // (callers can Set<string?>(k, null) → bytes "null" → Deserialize
            // returns null). Failure is reserved for actual parse errors.
            var value = JsonSerializer.Deserialize<T>(bytes, sr_options);
            return D2Result<T>.Ok(value);
        }
        catch (JsonException)
        {
            return new D2Result<T>(
                success: false,
                data: default,
                messages: [TK.Common.Errors.COULD_NOT_BE_DESERIALIZED],
                statusCode: System.Net.HttpStatusCode.InternalServerError,
                errorCode: ErrorCodes.COULD_NOT_BE_DESERIALIZED);
        }
        catch (NotSupportedException)
        {
            // STJ throws NotSupportedException for unsupported types (e.g.
            // some interface or polymorphic shape it can't construct).
            return new D2Result<T>(
                success: false,
                data: default,
                messages: [TK.Common.Errors.COULD_NOT_BE_DESERIALIZED],
                statusCode: System.Net.HttpStatusCode.InternalServerError,
                errorCode: ErrorCodes.COULD_NOT_BE_DESERIALIZED);
        }
    }
}

File: public/packages/typescript/caching/distributed-redis/src/json-cache-serializer.ts
// -----------------------------------------------------------------------
// SPDX-License-Identifier: Apache-2.0
// Copyright (c) DCSV
// -----------------------------------------------------------------------

import type { ICacheSerializer } from "@dcsv-io/d2-caching-abstractions";
import { TK } from "@dcsv-io/d2-i18n-keys";
import {
  ErrorCodes,
  fail,
  HttpStatusCode,
  ok,
  type D2Result,
} from "@dcsv-io/d2-result";

const TEXT_ENCODER = new TextEncoder();
const TEXT_DECODER = new TextDecoder();

/**
 * Default {@link ICacheSerializer} using `JSON.stringify` / `JSON.parse`.
 * Twin of .NET `JsonCacheSerializer` (STJ Web defaults): camelCase property
 * names, cycle graphs omit cyclic refs, no indent. Dev-friendly (Redis CLI
 * can inspect UTF-8 JSON values).
 */
export class JsonCacheSerializer implements ICacheSerializer {
  /** Stable format identifier for mixed-serializer archives. */
  readonly contentType = "application/json";

  /**
   * Serializes `value` to UTF-8 JSON bytes with camelCase property names
   * and cycle-safe omission of cyclic references.
   *
   * @returns `ok(bytes)` on success; `fail` with
   *   `ErrorCodes.COULD_NOT_BE_SERIALIZED` (HTTP 500) on encode failure.
   */
  serialize<T>(value: T): D2Result<Uint8Array> {
    try {
      const seen = new WeakSet<object>();
      const json = JSON.stringify(value, (_key, v: unknown) => {
        if (v !== null && typeof v === "object") {
          if (seen.has(v as object)) {
            return undefined;
          }

          seen.add(v as object);

          if (!Array.isArray(v)) {
            const out: Record<string, unknown> = {};

            for (const [k, prop] of Object.entries(
              v as Record<string, unknown>,
            )) {
              out[toCamelCase(k)] = prop;
            }

            return out;
          }
        }

        return v;
      });

      return ok(TEXT_ENCODER.encode(json));
    } catch {
      return fail<Uint8Array>({
        statusCode: HttpStatusCode.InternalServerError,
        errorCode: ErrorCodes.COULD_NOT_BE_SERIALIZED,
        messages: [TK.common.errors.COULD_NOT_BE_SERIALIZED],
      });
    }
  }

  /**
   * Deserializes UTF-8 JSON bytes. JSON `null` is a legitimate value.
   *
   * @returns `ok(value)` on success; `fail` with
   *   `ErrorCodes.COULD_NOT_BE_DESERIALIZED` (HTTP 500) on parse failure.
   */
  deserialize<T>(bytes: Uint8Array): D2Result<T> {
    try {
      if (bytes.byteLength === 0) {
        return fail<T>({
          statusCode: HttpStatusCode.InternalServerError,
          errorCode: ErrorCodes.COULD_NOT_BE_DESERIALIZED,
          messages: [TK.common.errors.COULD_NOT_BE_DESERIALIZED],
        });
      }

      const text = TEXT_DECODER.decode(bytes);
      const value = JSON.parse(text) as T;

      return ok(value);
    } catch {
      return fail<T>({
        statusCode: HttpStatusCode.InternalServerError,
        errorCode: ErrorCodes.COULD_NOT_BE_DESERIALIZED,
        messages: [TK.common.errors.COULD_NOT_BE_DESERIALIZED],
      });
    }
  }
}

function toCamelCase(name: string): string {
  const first = name[0];

  if (first === undefined) {
    return name;
  }

  if (first === first.toLowerCase()) {
    return name;
  }

  return first.toLowerCase() + name.slice(1);
}
